core/src/main/java/ai/timefold/solver/core/impl/domain/variable/declarative/GraphStructure.java

```diff
@@ -0,0 +1,108 @@
+package ai.timefold.solver.core.impl.domain.variable.declarative;
+
+import java.util.Arrays;
+
+import ai.timefold.solver.core.impl.domain.solution.descriptor.SolutionDescriptor;
+import ai.timefold.solver.core.impl.util.MutableInt;
+import ai.timefold.solver.core.preview.api.domain.metamodel.VariableMetaModel;
+
+import org.jspecify.annotations.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public enum GraphStructure {
+    /**
+     * A graph structure that only accepts the empty graph.
+     */
+    EMPTY,
+
+    /**
+     * A graph structure without dynamic edges. The topological order
+     * of such a graph is fixed, since edges are neither added nor removed.
+     */
+    NO_DYNAMIC_EDGES,
+
+    /**
+     * A graph structure where there is at most
+     * one directional parent for each graph node, and
+     * no indirect parents.
+     * For example, when the only input variable from
+     * a different entity is previous. This allows us
+     * to use a successor function to find affected entities.
+     * Since there is at most a single parent node, such a graph
+     * cannot be looped.
+     */
+    SINGLE_DIRECTIONAL_PARENT,
+
+    /**
+     * A graph structure that accepts all graphs.
+     */
+    ARBITRARY;
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(GraphStructure.class);
+
+    public record GraphStructureAndDirection(GraphStructure structure,
+            @Nullable VariableMetaModel<?, ?, ?> parentMetaModel,
```

Comment:
If using `Nullable`, please mark the class as `NullMarked`.
This will make everything not annotated as `Nullable` explicitly `NonNullable`.
If not using `NullMarked`, `structure` nullness is undefined, while the other two arguments have it defined - this is inconsistent and confusing.